algorithms/nautilus/fuzzer/state.cpp

```diff
@@ -0,0 +1,469 @@
+          break;
+      }
+      break;
+    }
+
+    case PUTExitReasonType::FAULT_TMOUT: { /* Timeout */
+      /* Get current datetime */
+      const std::time_t t = std::time(nullptr);
+      const std::tm* tm = std::localtime(&t);
+
+      /* Update last timeout */
+      std::ostringstream oss;
+      oss << std::put_time(tm, "[%Y-%m-%d] %H:%M:%S");
+      // TODO: Use lock when multi-threaded
+      last_timeout = oss.str();
+      total_found_hang++;
+
+      /* Stringify tree */
+      std::string buffer;
+      tree_like.UnparseTo(ctx, buffer);
+
+      /* Save tree to file */
+      std::string filepath = Util::StrPrintf(
+        "%s/timeout/%09ld",
+        setting->path_to_workdir.c_str(), execution_count
+      );
+      int fd = Util::OpenFile(filepath,
+                              O_WRONLY | O_CREAT | O_TRUNC,
+                              S_IWUSR | S_IRUSR); // 0600
+      if (fd == -1) {
+        /* Print testcase because we don't want to lose it */
+        std::cout << buffer << std::endl;
+        throw exceptions::unable_to_create_file(
+          Util::StrPrintf("Cannot save timeout: %s", filepath.c_str()),
+          __FILE__, __LINE__
+        );
+      }
+      Util::WriteFile(fd, buffer.data(), buffer.size());
+      Util::CloseFile(fd);
+
+      break;
+    }
+
+    case PUTExitReasonType::FAULT_CRASH: { /* Signal */
+      if (new_bits.size() == 0) break;
+
+      /* Get current datetime */
+      const std::time_t t = std::time(nullptr);
+      const std::tm* tm = std::localtime(&t);
+
+      /* Update last sig */
+      std::ostringstream oss;
+      oss << std::put_time(tm, "[%Y-%m-%d] %H:%M:%S");
+      // TODO: Use lock when multi-threaded
+      total_found_sig++;
+      // TODO: Use lock when multi-threaded
+      last_found_sig = oss.str();
+
+      /* Stringify tree */
+      std::string buffer;
+      tree_like.UnparseTo(ctx, buffer);
+
+      /* Save tree to file */
+      std::string filepath = Util::StrPrintf(
+        "%s/signaled/%d_%09ld",
+        setting->path_to_workdir.c_str(), exit_status.signal, execution_count
+      );
+      int fd = Util::OpenFile(filepath,
+                              O_WRONLY | O_CREAT | O_TRUNC,
+                              S_IWUSR | S_IRUSR); // 0600
+      if (fd == -1) {
+        /* Print testcase because we don't want to lose it */
+        std::cout << buffer << std::endl;
+        throw exceptions::unable_to_create_file(
+          Util::StrPrintf("Cannot save crash: %s", filepath.c_str()),
+          __FILE__, __LINE__
+        );
+      }
+      Util::WriteFile(fd, buffer.data(), buffer.size());
+      Util::CloseFile(fd);
+
+      break;
+    }
+
+    default: // pass
+      break;
+  }
+  // exit_reason
+  exec_reason = exec_reason;
+}
+
+/**
+ * @fn
+ * @brief Execute PUT with raw data as input
+ * @param (code) Testcase
+ * @return Execution time in nanoseconds
+ */
+u64 NautilusState::ExecRaw(const std::string& code) {
```

Comment:
This can bring a big change, so I'm hesitating to point out it, but I think holding inputs for PUT in `std::string` is not a good idea because `std::string` is (originally) meant to hold (printable) strings, not byte sequences. Though std::string has a guarantee of having contiguous memory now, `std::vector` is still better and safer way to hold them. https://stackoverflow.com/questions/9626914/stdstring-or-stdvectorchar-to-hold-raw-data

Reply:
I decided to use `std::string` instead of `std::vector<u8>` mainly because those data usually hold printable string, such as grammar and inputs. In fact, there exists bunch of codes to convert `str` to `Vec<u8>` (vice versa) especially in the test cases and it was troublesome to implement it in C++.

Reply:
While that decision is somewhat reasonable and definitely one of the possible choices, binary file formats also can have grammar and it sounds a bit weak as the main reason. 
I would agree to use `std::string` if unparsing becomes really painstaking by adopting `std::vector` (for some reason like too much operations on strings, concatnation, search, etc.).
But replacing `std::string` with `std::vector` looks almost straightforward; it would be almost done by rewriting this function (correct me if I'm mistaken).
https://github.com/fuzzuf/fuzzuf/pull/23/files#diff-ede30e90846b5b2593334822605eadd82fef6a24b1864c680674e4c9cda9b211R536-R542
Also, the original implementation also adopts `vec<u8>`, not `String`.

Considering these situation, I want to hear your opinion.
If you agree with me, but fix you don't have time to fix, then leaving comments is enough.

Reply:
It's not trivial to convert some of the codes from Rust, such as `std::io::Cursor`, which is currently altered by `std::stringstream`. Also, I need to re-write the rule of grammar in every test case into something like `{'a', 'b', '{', 'C', '}'}` (which is obviously dirty), as well as changing the implementation of the grammar loader.
I haven't faced any troubles on using `std::string` as byte array as far as I've tested. It can handle null bytes, whitespaces, and newlines as intended.

So, I want to leave it as is for now. Or, maybe we can alias it to something like `bytes` to prevent confusion.

Reply:
FYI, `std::stringstream` can be used with other data structures than `std::string,` so I think that's not so problematic.
The second reason sounds also a bit weak. Can't it be dealt with by creating a helper function of converting `std::string` to `std::vector`?
If you need to use the helper function in a fuzzer itself and it affects the performance, then you can legitimately avoid it, but I think you can't if the function is used only in tests.

But only the thing I'm worried about is the NULL byte, so if you tested it, then probably it's no problem.
Still, I'm not fully persuaded and guess there are not enough tests to insist it's no problem to use std::string, so comments about it should be left.

Reply:
Binary data is now supported in [0802f59](https://github.com/fuzzuf/fuzzuf/pull/23/commits/0802f59d485ea9fab618d99b3462eaa5332a47ae)
I will leave it as TODO ([84b24a8](https://github.com/fuzzuf/fuzzuf/pull/23/commits/84b24a866646b06e63c9910a7e1225cdf6523fe6))

Reply:
Wonderful